## Task 1

In [10]:
import pandas as pd
import numpy as np
df = pd.read_csv("bangalore_tech_salaries (1).csv")
df.head()
print(df.shape)
print(df.columns)
df.info()
print(df.dtypes)
print(df.isnull().sum())
print(df.duplicated().sum())
df.describe()

(1015, 12)
Index(['Employee_ID', 'Role ', 'years_exp', 'Current_CTC', 'Previous_CTC',
       'Company', 'company_TYPE', 'Skills', 'Location', 'Education_Tier',
       'Joining_Year', 'Work_Mode'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Employee_ID     1015 non-null   object
 1   Role            1015 non-null   object
 2   years_exp       1015 non-null   int64 
 3   Current_CTC     1015 non-null   object
 4   Previous_CTC    815 non-null    object
 5   Company         1015 non-null   object
 6   company_TYPE    1015 non-null   object
 7   Skills          988 non-null    object
 8   Location        995 non-null    object
 9   Education_Tier  1015 non-null   object
 10  Joining_Year    1015 non-null   int64 
 11  Work_Mode       1015 non-null   object
dtypes: int64(2), object(10)
memory usage: 95.3+ KB
Employ

,years_exp,Joining_Year
count,1015.000000,1015.000000
mean,3.429557,2022.698522
std,2.558721,1.308351
min,0.000000,2020.000000
25%,1.000000,2022.000000
50%,3.000000,2023.000000
75%,5.000000,2024.000000
max,12.000000,2024.000000


## Task 2: Data Cleaning

In [24]:
# Convert column names into snake_case format
df.columns = (
    df.columns
      .str.strip()
      .str.lower()
      .str.replace(" ", "_")
)
print(df.columns)
# Display all unique role names
print(df["role"].unique())
print(df["company_type"].unique())
print(df["education_tier"].unique())
print(df["work_mode"].unique())

# Create a cleaned role column
role_mapping = {
    "Backend Dev": "SDE Backend",
    "Backend Developer": "SDE Backend",
    "Backend Engineer": "SDE Backend",
    "BE": "SDE Backend",
    "SDE-Backend": "SDE Backend",

    "Frontend Dev": "SDE Frontend",
    "Frontend Developer": "SDE Frontend",
    "Frontend Engineer": "SDE Frontend",
    "FE": "SDE Frontend",
    "SDE-Frontend": "SDE Frontend",

    "FullStack": "SDE Full-Stack",
    "Fullstack Dev": "SDE Full-Stack",
    "Full Stack Engineer": "SDE Full-Stack",
    "SDE FS": "SDE Full-Stack",

    "DS": "Data Scientist",
    "Data Science Engineer": "Data Scientist",
    "ML Engineer": "Data Scientist",

    "DA": "Data Analyst",
    "BI Analyst": "Data Analyst",
    "Analytics Engineer": "Data Analyst",

    "BA": "Business Analyst",
    "Business Systems Analyst": "Business Analyst",

    "PM": "Product Manager",
    "Sr PM": "Product Manager",
    "Product Lead": "Product Manager",

    "UI Designer": "UI/UX",
    "UX Designer": "UI/UX",
    "Designer": "UI/UX",
    "Product Designer": "UI/UX",

    "DevOps": "DevOps Engineer",

    "Infra Engineer": "Site Reliability Engineer",
    "SRE": "Site Reliability Engineer"
}

df["role_clean"] = df["role"].replace(role_mapping)
print(df["role_clean"].value_counts())

# Standardize company type names
company_mapping = {
    "mnc": "MNC",
    "unicorn": "Unicorn",
    "UNICORN": "Unicorn",
    "mid-size": "Mid-size",
    "MID-SIZE": "Mid-size",
    "early-stage": "Early-stage",
    "EARLY-STAGE": "Early-stage"
}

df["company_type_clean"] = df["company_type"].replace(company_mapping)
print(df["company_type_clean"].value_counts())

# Standardize education tier values
education_mapping = {
    "Tier-1": "Tier 1",
    "T1": "Tier 1",
    "1": "Tier 1",

    "Tier-2": "Tier 2",
    "T2": "Tier 2",
    "2": "Tier 2",

    "Tier-3": "Tier 3",
    "T3": "Tier 3",
    "3": "Tier 3"
}

df["education_tier_clean"] = df["education_tier"].replace(education_mapping)
print(df["education_tier_clean"].value_counts())

# Standardize work mode values
work_mode_mapping = {
    "WFH": "Remote",
    "Work from Office": "Onsite",
    "Hybrid (3 days)": "Hybrid"
}

df["work_mode_clean"] = df["work_mode"].replace(work_mode_mapping)
print(df["work_mode_clean"].value_counts())

# Check different salary formats present in Current_CTC
print(df["current_ctc"].head(20))
print(df["previous_ctc"].head(20))

# Convert different salary formats into LPA (float)
def parse_ctc(value):

    # Handle missing values
    if pd.isna(value):
        return np.nan

    # Convert everything to string
    value = str(value).strip()

    # Remove currency symbol and LPA text
    value = value.replace("₹", "")
    value = value.replace("LPA", "")
    value = value.replace("lpa", "")
    value = value.strip()

    # Case 1: Salary stored as 15,50,000
    if "," in value:
        salary = float(value.replace(",", ""))

        # Convert rupees into LPA
        if salary > 100000:
            return salary / 100000

    # Case 2: Already stored as 15.5
    return float(value)
df["current_ctc_lpa"] = df["current_ctc"].apply(parse_ctc)

df["previous_ctc_lpa"] = df["previous_ctc"].apply(parse_ctc)

df[["current_ctc", "current_ctc_lpa"]].head(15)

print(df[["current_ctc_lpa", "previous_ctc_lpa"]].isnull().sum())

df = df.drop_duplicates()

print("Dataset shape after removing duplicates:", df.shape)

print(df.dtypes)
print(df["role_clean"].value_counts())
print(df["company_type_clean"].value_counts())
print(df["education_tier_clean"].value_counts())
df[["current_ctc", "current_ctc_lpa"]].head(10)

Index(['employee_id', 'role', 'years_exp', 'current_ctc', 'previous_ctc',
       'company', 'company_type', 'skills', 'location', 'education_tier',
       'joining_year', 'work_mode', 'role_clean', 'company_type_clean',
       'education_tier_clean', 'work_mode_clean', 'current_ctc_lpa',
       'previous_ctc_lpa'],
      dtype='object')
['DS' 'DevOps' 'SDE Backend' 'SDE Frontend' 'Product Lead'
 'Product Manager' 'Site Reliability Engineer' 'DA' 'UI/UX' 'SDE-Frontend'
 'UI Designer' 'ML Engineer' 'Frontend Dev' 'Data Scientist'
 'Full Stack Engineer' 'Designer' 'Product Designer' 'BA'
 'Frontend Engineer' 'BI Analyst' 'Backend Dev' 'Sr PM' 'FE' 'SDE-Backend'
 'Data Analyst' 'Business Analyst' 'Backend Developer'
 'Business Systems Analyst' 'Data Science Engineer' 'FullStack' 'BE'
 'Frontend Developer' 'Backend Engineer' 'UX Designer' 'Infra Engineer'
 'PM' 'SDE FS' 'DevOps Engineer' 'Analytics Engineer' 'Fullstack Dev'
 'SDE Full-Stack' 'SRE']
['Unicorn' 'MNC' 'Mid-size' 'early-stage' 

,current_ctc,current_ctc_lpa
0,49.4,49.4
1,9.7,9.7
2,"3,360,000",33.6
3,41.4 LPA,41.4
4,"3,640,000",36.4
5,32.1,32.1
6,24.6,24.6
7,21.7 LPA,21.7
8,21.0,21.0
9,26.9 LPA,26.9


## Task 3.1: CTC Distribution by Role

In [26]:
# Calculate salary statistics for each cleaned role

role_ctc = (
    df.groupby("role_clean")["current_ctc_lpa"]
      .agg(
          Median_CTC="median",
          Mean_CTC="mean",
          Min_CTC="min",
          Max_CTC="max"
      )
      .sort_values(by="Median_CTC", ascending=False)
)

print(role_ctc)
Median_CTC="median"
Mean_CTC="mean"
Min_CTC="min"
Max_CTC="max"

                           Median_CTC   Mean_CTC  Min_CTC  Max_CTC
role_clean                                                        
Product Manager                 31.30  34.131250     10.8     80.1
Data Scientist                  24.70  27.579200     10.0     75.6
Site Reliability Engineer       23.30  23.481481      8.8     55.0
SDE Full-Stack                  22.35  25.409375      8.9     71.7
SDE Frontend                    21.20  22.193333      6.7     84.4
SDE Backend                     21.10  23.284210      7.9     55.1
Business Analyst                19.90  21.873148      6.8     52.7
DevOps Engineer                 19.45  22.891304      9.7     60.3
UI/UX                           18.90  20.968142      6.2     63.3
Data Analyst                    16.90  17.874000      5.2     43.4


### Insight (Q3.1)

- Product Manager has the highest median salary at **31.30 LPA**, making it the highest-paying role in this dataset.
- Data Analyst has the lowest median salary at **16.90 LPA**, which is significantly lower than most engineering roles.
- Roles such as Product Manager and SDE Full-Stack show a noticeable gap between the mean and median CTC, suggesting the presence of high-salary outliers.

##Task 3.2 – Experience Curve for SDE Backend

In [37]:
# Filter only SDE Backend employees
backend_df = df[df["role_clean"] == "SDE Backend"].copy()

# Create experience groups
bins = [-1, 1, 3, 5, 100]
labels = ["0-1 Years", "2-3 Years", "4-5 Years", "6+ Years"]

backend_df["exp_band"] = pd.cut(
    backend_df["years_exp"],
    bins=bins,
    labels=labels
)

# Calculate median CTC for each experience band
experience_ctc = (
    backend_df
    .groupby("exp_band", observed=False)["current_ctc_lpa"]
    .median()
    .reset_index()
)

# Calculate percentage growth
experience_ctc["Growth_%"] = (
    experience_ctc["current_ctc_lpa"]
    .pct_change() * 100
)

print(experience_ctc)

    exp_band  current_ctc_lpa   Growth_%
0  0-1 Years            11.65        NaN
1  2-3 Years            20.00  71.673820
2  4-5 Years            25.85  29.250000
3   6+ Years            40.40  56.286267


### Insight (Q3.2)

- The median salary for SDE Backend professionals increases from **11.65 LPA** (0–1 years) to **40.40 LPA** (6+ years).
- The highest salary growth (**71.67%**) occurs between **0–1 years** and **2–3 years**, indicating rapid early-career progression.
- Professionals who build strong technical skills and gain experience during their first three years can significantly improve their earning potential.

##Task 3.3 – Skill Premium for SDEs

In [49]:
# Filter only SDE roles
sde_df = df[df["role_clean"].isin([
    "SDE Backend",
    "SDE Frontend",
    "SDE Full-Stack"
])].copy()

# Skills to compare
skills = ["AWS", "ML", "System Design", "Kubernetes"]

# Store results
results = []

# Compare salary for employees with and without each skill
for skill in skills:

    with_skill = sde_df[
        sde_df["skills"].str.contains(skill, case=False, na=False)
    ]

    without_skill = sde_df[
        ~sde_df["skills"].str.contains(skill, case=False, na=False)
    ]

    median_with = with_skill["current_ctc_lpa"].median()
    median_without = without_skill["current_ctc_lpa"].median()

    premium = ((median_with - median_without) / median_without) * 100

    results.append([
        skill,
        round(median_with, 2),
        round(median_without, 2),
        round(premium, 2)
    ])

# Create DataFrame
skill_premium = pd.DataFrame(
    results,
    columns=[
        "Skill",
        "Median With Skill",
        "Median Without Skill",
        "Premium (%)"
    ]
)

# Sort by premium
skill_premium = skill_premium.sort_values(
    by="Premium (%)",
    ascending=False
)

print(skill_premium)

           Skill  Median With Skill  Median Without Skill  Premium (%)
2  System Design              25.30                  21.0        20.48
1             ML              23.95                  21.3        12.44
3     Kubernetes              23.20                  21.5         7.91
0            AWS              20.95                  21.9        -4.34


### Insight (Q3.3)

- Among the selected skills, **System Design** provides the highest salary premium, with professionals earning **20.48%** more than those without this skill.
- **Machine Learning (ML)** also offers a noticeable salary advantage of **12.44%**, while **Kubernetes** provides a moderate premium of **7.91%**.
- Interestingly, professionals with **AWS** have a **4.34% lower** median salary than those without AWS in this dataset, indicating that AWS alone may not be a strong salary differentiator and is often expected as a common skill.

## Task 3.4 – Company Type Premium

In [53]:
# Filter only SDE Backend employees
backend_df = df[df["role_clean"] == "SDE Backend"].copy()

# Calculate median salary by company type
company_ctc = (
    backend_df
    .groupby("company_type_clean")["current_ctc_lpa"]
    .median()
    .reset_index()
)

print(company_ctc)

# Get median salaries
unicorn_salary = company_ctc.loc[
    company_ctc["company_type_clean"] == "Unicorn",
    "current_ctc_lpa"
].values[0]

mnc_salary = company_ctc.loc[
    company_ctc["company_type_clean"] == "MNC",
    "current_ctc_lpa"
].values[0]

# Calculate percentage premium
premium = ((unicorn_salary - mnc_salary) / mnc_salary) * 100

print(f"Unicorn Premium over MNC: {premium:.2f}%")
unicorn_salary = 26
mnc_salary = 20

  company_type_clean  current_ctc_lpa
0        Early-stage            18.60
1                MNC            20.45
2           Mid-size            19.50
3            Unicorn            27.35
Unicorn Premium over MNC: 33.74%


### Insight (Q3.4)

- Among SDE Backend professionals, **Unicorn companies offer the highest median salary of 27.35 LPA**, while **Early-stage companies offer the lowest median salary of 18.60 LPA**.
- The median salary in **Unicorn companies is 33.74% higher than in MNCs**, indicating a significant company-type premium for the same role.
- Candidates seeking higher compensation may benefit from targeting Unicorn companies, while also considering factors such as job stability, learning opportunities, and work-life balance.

##Task 3.5 – Top 10 Most Underpaid Professionals

In [56]:
# Create experience bands for the complete dataset

bins = [-1, 1, 3, 5, 100]

labels = ["0-1 Years", "2-3 Years", "4-5 Years", "6+ Years"]

df["exp_band"] = pd.cut(
    df["years_exp"],
    bins=bins,
    labels=labels
)
# Count employees in each group

group_size = (
    df.groupby(
        ["role_clean", "company_type_clean", "exp_band"],
        observed=False
    )["employee_id"]
    .transform("count")
)

df["group_size"] = group_size
# Calculate median salary within each group

df["group_median"] = (
    df.groupby(
        ["role_clean", "company_type_clean", "exp_band"],
        observed=False
    )["current_ctc_lpa"]
    .transform("median")
)
# Difference between employee salary and group median

df["gap"] = (
    df["current_ctc_lpa"] -
    df["group_median"]
)
# Keep groups with at least 10 employees

underpaid_df = df[
    df["group_size"] >= 10
]
# Sort employees by salary gap

underpaid_df = underpaid_df.sort_values(
    by="gap"
)
# Display top 10 most underpaid professionals

top10 = underpaid_df[
    [
        "employee_id",
        "role_clean",
        "company_type_clean",
        "years_exp",
        "current_ctc_lpa",
        "group_median",
        "gap"
    ]
].head(10)

print(top10)


    employee_id        role_clean company_type_clean  years_exp  \
810     BLR0615    Data Scientist           Mid-size          6   
460     BLR0141    Data Scientist           Mid-size          6   
904     BLR0925       SDE Backend                MNC          4   
250     BLR0072       SDE Backend                MNC          4   
249     BLR0733  Business Analyst                MNC          7   
728     BLR0301   Product Manager                MNC          2   
386     BLR0453  Business Analyst                MNC          6   
753     BLR0611   Product Manager                MNC          2   
688     BLR0203       SDE Backend                MNC          4   
711     BLR0451      SDE Frontend                MNC          4   

     current_ctc_lpa  group_median    gap  
810             28.1         40.30 -12.20  
460             31.7         40.30  -8.60  
904             19.4         27.30  -7.90  
250             19.5         27.30  -7.80  
249             31.1         38.50  -7.40 

### Insight (Q3.5)

- The most underpaid employee in the dataset is **BLR0615 (Data Scientist)**, earning **28.10 LPA**, which is **12.20 LPA below** the median salary of professionals with the same role, company type, and experience level.
- Several employees have salaries significantly below their respective group medians, indicating noticeable compensation differences even among professionals with similar roles, experience, and company types.
- This analysis helps identify employees who may be underpaid relative to their peers and can support salary benchmarking and compensation review decisions.

## Task 4 – Final Summary Report

In [57]:
# Create summary report

summary_report = pd.DataFrame({

    "Metric": [
        "Total Employees",
        "Unique Roles",
        "Average Current CTC",
        "Median Current CTC",
        "Highest Paying Role",
        "Lowest Paying Role",
        "Highest Paying Company Type"
    ],

    "Value": [
        len(df),
        df["role_clean"].nunique(),
        round(df["current_ctc_lpa"].mean(), 2),
        round(df["current_ctc_lpa"].median(), 2),
        role_ctc.index[0],
        role_ctc.index[-1],
        company_ctc.sort_values(
            by="current_ctc_lpa",
            ascending=False
        )["company_type_clean"].iloc[0]
    ]

})

summary_report

,Metric,Value
0,Total Employees,1000
1,Unique Roles,10
2,Average Current CTC,24.17
3,Median Current CTC,21.8
4,Highest Paying Role,Product Manager
5,Lowest Paying Role,Data Analyst
6,Highest Paying Company Type,Unicorn


### Final Report Summary

- The dataset contains information for **1015 employees** across **10 standardized job roles**.
- **Product Manager** is the highest-paying role based on median salary, while **Data Analyst** has the lowest median salary.
- **Unicorn companies** offer the highest median compensation among all company types.
- Salary generally increases with experience, especially during the early years of a professional's career.
- Skills such as **System Design** and **Machine Learning** are associated with higher salaries among SDE professionals in this dataset.

## Task 5 – Final Conclusions

### Key Takeaways

1. Salary growth is strongly influenced by experience, particularly during the first few years of a software engineering career.

2. Product Managers receive the highest median salaries, whereas Data Analysts have the lowest median salaries in this dataset.

3. Unicorn companies provide significantly higher compensation than other company types for similar roles.

4. Among the selected technical skills, System Design offers the highest salary premium, highlighting its importance for career growth.

5. Salary benchmarking helped identify professionals who are paid below the median of their peer group, which can support fair compensation analysis.

# Employee Salary Analysis Project
### Live Project Submission
**Name:** Your Name

**Date:** 28 June 2026

Employee Salary Analysis Project

1. Import Libraries

2. Load Dataset

3. Initial Data Inspection

4. Task 2 - Data Cleaning

5. Task 3.1 - CTC Distribution by Role

6. Task 3.2 - Experience Curve

7. Task 3.3 - Skill Premium

8. Task 3.4 - Company Type Premium

9. Task 3.5 - Underpaid Professionals

10. Task 4 - Final Summary Report

11. Task 5 - Final Conclusions

In [65]:
# Create a summary report containing key project findings

summary_report = pd.DataFrame({

    "Metric": [
        "Total Employees",
        "Unique Roles",
        "Average Current CTC (LPA)",
        "Median Current CTC (LPA)",
        "Highest Paying Role",
        "Lowest Paying Role",
        "Highest Paying Company Type"
    ],

    "Value": [
        len(df),
        df["role_clean"].nunique(),
        round(df["current_ctc_lpa"].mean(), 2),
        round(df["current_ctc_lpa"].median(), 2),
        role_ctc.index[0],
        role_ctc.index[-1],
        company_ctc.loc[
            company_ctc["current_ctc_lpa"].idxmax(),
            "company_type_clean"
        ]
    ]

})

print(summary_report)

                        Metric            Value
0              Total Employees             1000
1                 Unique Roles               10
2    Average Current CTC (LPA)            24.17
3     Median Current CTC (LPA)             21.8
4          Highest Paying Role  Product Manager
5           Lowest Paying Role     Data Analyst
6  Highest Paying Company Type          Unicorn


### Insight (Task 6 – Final Summary)

- The employee salary dataset was successfully cleaned, standardized, and analyzed using Python and Pandas.
- Product Manager emerged as the highest-paying role, while Data Analyst had the lowest median salary in the dataset.
- Salary generally increased with experience, with the highest growth observed during the early years of a professional's career.
- Among the selected technical skills, System Design showed the highest salary premium, indicating its strong influence on compensation.
- Unicorn companies offered the highest median salaries compared to MNCs, Mid-size, and Early-stage companies.
- Salary benchmarking also identified employees earning significantly below the median of their peer group, providing useful insights for compensation analysis.
- Overall, the project demonstrates how data preprocessing, exploratory data analysis, and business-focused insights can support better salary benchmarking and informed career or organizational decisions.